doc/designs/evp_skey.md

```diff
@@ -0,0 +1,157 @@
+EVP_KEYMGMT object and an opaque pointer to some internal representation.
+EVP_KEYMGMT specifies what parameters are acceptable by the chosen provider and
+may be specific algorithms of this provider. These parameters are represented
+as an array of OSSL_PARAM.
+
+The way we have used OSSL_PARAM is for things that are (or should be) unknown
+from a libcrypto perspective, while we have made function arguments from what
+is known.
+
+Among the things that are known is that all ciphers take a key,
+unconditionally. With that in mind, it's logical that it has become a function
+argument. At the time, we didn't think that a key would ever be anything but "a
+bunch of bytes", which is why it's an `unsigned char *`. IV is also a separate
+argument for historical reasons.
+
+There are other parameters that some but not all ciphers take. You'll find a
+substantial list in the manual for EVP_CipherInit(), and they are suitable for
+OSSL_PARAM. The difference here is that *those* params are not
+provider-specific but algorithm-specific.
+
+Things that were previously passed to backend implementations with ctrl
+functions have become EVP_CIPHER_CTX level OSSL_PARAM items with providers.
+
+Provider perspective
+--------------------
+
+On the provider end, encryption and decryption implementations would receive an
+opaque pointer similar to the provkey pointer used with asymmetric operations.
+They expected to know how to deal with the structure. This pointer is a result
+of processing an array of OSSL_PARAM by the EVP_KEYMGMT import operation.
+
+Once created by import, the EVP_SKEY structure can be exported to the extent
+the provider and EVP_KEYMGMT suports it. There is no guarantee that the result
+of export-import operation is the same as the initial set of the parameters.
+
+The EVP_SKEY object *can* have an algorithm association. It should be a
+provider-specific decision or maybe an information that could be associated to
+an EVP_SKEY any time later. Whether it starts with being algorithm specific or
+not depends on how it was created. If it's a result of obtaining a PKCS#11
+secret key, then it will most likely be algorithm specific from the start.
+
+We do not see why we would want to assign an algorithm immediately during
+the creation of an EVP_SKEY as result of derive() operation, if the operation
+just creates bunch of bytes that can be used as a key equally well for HMAC or
+AES encryption.
+
+Objects relationships
+---------------------
+
+We have EVP_CIPHER object associated with a particular provider. A EVP_SKEY
+object should be compatible with the particular EVP_CIPHER object. After that
+it is used within a EVP_CIPHER_CTX object associated with the same EVP_CIPHER
+object.
+
+The way to reuse the key within the other provider is through the export/import
+dance. As mentioned before, it may fail. The way to reuse it with a different
+cipher context is up_ref (we don't support copying).
```

Comment:
This means, modifying one duplicated EVP_SKEY also duplicate all other instances? What about hardware-backed keys, which allow real key duplication? Is there a chance to support one of the behaviours (per provider?), real key duplication and up_ref?

Reply:
I think this text does not reflect the code changes.
That said there are 2 cases here:
1) Key on HSM, in that case the parameters will carry something like a pkcs11 URI and there is no key duplication, you just get references to the same key.
2) actual key buffer, in that case you will import the key on a HSM session (if possible) and remove it from the HSM once the EVP_SKEY is freed.

I do not think there is any case where openssl should ever attempt a key dup, and in fact I would ask that a key duplication API is not even considered, any application written to use such an API would fail to use HW modules in most cases.

Reply:
@t8m there are some cases, where a key object can not be duplicated, e.g. in pkcs#11, if the CKA_COPYABLE attribute of a key object is false. In this case, the function should return with an error. 

But in general, there is no reason why a provider should not copy a key in EVP_SKEY_dup().

Reply:
@holger-dengler I don't think we should permit modifications of EVP_SKEY objects after creation.

Reply:
@t8m we can allow key duplication if the keymgmt allows it - but I currently don't see viable scenarios to do it.

Reply:
@eldmit for the hardware-backed keys on s390x we have use cases for key duplication. It would be nice, if this is supported by EVP_SKEY.

Reply:
Implemented and it is up to the provider whether it supports this option

Reply:
@holger-dengler can you clarify those use cases? How does it work?

Reply:
Doesn't matter really, I've already added this callback - and it is not mandatory-to-implement

Reply:
The implementation of `EVP_SKEY_dup()` looks sufficient for our use-case. I'll do a quick protecrted key prototype in the next few days based on the fake-provider. But I'm sure, that I won't miss anything.

Reply:
> @holger-dengler can you clarify those use cases? How does it work?

@kroeckx the protected keys are symmetric keys, which are wrapped with a machine key. The s390 architecture provides instructions to operate with these wrapped (protected) keys. The protected key blobs are bigger than clear keys of the same type, because they also contain a kind of verification information for the machine. Protected keys can be created based on clear keys and - more important - from HSM key material. The main benefit in the latter case is, that operations with the protected keys are processed CPU-local without further IO to the HSM.

Key duplication is - more or less - a copy of the protected key blob (+ machine verification).

Reply:
@holger-dengler I still don't understand why do you need a duplication instead of just the up_ref

Reply:
So I think @holger-dengler is talking about importing from HSM to a wrapped protected key. That doesn't look like a duplicate operation to me. A dup function would:
- For a key in the HSM: Have 2 copies of the same key in the HSM. We tell the HSM to copy it to a new location.
- For a wrapped protected key: Have 2 identical copies wrapped protected keys in RAM. As in, we copy all the wrapped bytes to a new location.
- For a clear/unprotected key: Have 2 identical copies clear/unprotected keys in RAM.

None of those options seem useful to me. What you might need is a way to import/convert between the different types. A provider could just import it from the HSM into a wrapped protected key. We don't want to have to change the application software to be able to do that.

Reply:
they kinda are

Reply:
Checking users of EVP_PKEY_dup(), it actually doesn't seems used much. We should probably convert internal use of it, and think about deprecating it.